### DISCUSSION TITLE
Wildcard Indexing API Proposal

### DISCUSSION THREAD

**Discussion by JonoPrest** [enviodev/hyperindex#123] [General]
Wildcard config will be a second optional argument in handlers
```ts
ERC20.Transfer.handler(
  async ({ event, context }) => {
    // ...
  },
  //options
  { wildcard: true, filter: { from: ["0x123"], to: ["0x345", "0x789"] } },
);
```
1. This allows us to provide types for filters (no need to interperate topic0, topic1 etc. and create hexadecimal values for them respectively)
2. It aligns with the goal of having less reliance on codegen from configuration
3. It removes the need for running a codegen step and keeps the dev experience in one place

Spec:

All `options` object fields are optional. Wildcard field defaults to false.

If all networks related to this event are hypersync, you can pass multiple filters. (RPC does not support multiple topic selections)
```ts
{ wildcard: false, filter: [{ from: ["0x123"] }, { to: ["0x345"] }] }
```

In the case where no filters are passed and wildcard is set to true, that means indexing all events with the matching topic0 on the network
```ts
{ wildcard: true }
```

In the case where multiple contracts have the same event configured, wildcard option cannot be set to true as it would be difficult to define a sensible behaviour for which handler a given matching event should be routed to.

Known limitations:

1. You cannot index based on topics 1-3 without topic 0
2. You cannot index multiple wildcard filters on the same event

**Comment by moose-code** [comment]
Remove certain spam NFT collections etc:
Negative filters for wildcards (excl. certain topics)

**Reply by JasoonS** [reply]
I guess that would be a negative fillter on the address, not on the topic. But make sense for topics too. Can you think of a use case for negative topic filters?

I think negative filters should be implemented client side since there is no analog in hypersync.

I think this would be an additional feature and not part of the first release of wildcard filters.

**Reply by JonoPrest** [reply]
It also doesn't exist in RPC filtering so it would definitely be a special case.

**Reply by moose-code** [reply]
Got you makes sense 👍  Was relaying thoughts from Krauser in a meeting last week.
